File: CompilerCM/src/token/TokenList.java
package scanner.token;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class TokenList implements ITokenList {
	private List<IToken> tokens;
	private Iterator<IToken> iterator;

	public TokenList() {
		tokens = new LinkedList<>();
	}

	@Override
	public void add(IToken token) {
		tokens.add(token);

	}

	@Override
	public void reset() {
		tokens.clear();
		iterator = null;
	}

	@Override
	public IToken nextToken() {
		if (iterator == null)
			iterator = tokens.iterator();
		return iterator.next();
	}
	
	@Override
	public String toString(){
		String output ="[";
		for( IToken t : tokens){
			System.out.println(t.toString());
			output += t.toString()+", ";
		}
		return output+"]";
	}

}

File: CompilerMR/IML_Compiler/Scanner/DataTypes/TokenList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace IML_Compiler.IMLScanner.DataTypes
{
    class TokenList : ITokenList
    {
        private List<Token> tokens;
        private int pos;

        public TokenList()
        {
            tokens = new List<Token>();
            pos = 0;
        }


        public void Add(Token token)
        {
            tokens.Add(token);
        }

        public void Reset()
        {
            pos = 0;
        }

        public Token NextToken()
        {
            if (pos >= tokens.Count) return null;

            return tokens[pos++];
        }

        public override string ToString()
        {
            var str = tokens.Select(a => a.ToString()).Aggregate((a, b) => a + ", " + b);
            return $"[{str}]";
        }
    }
}
